Minor fix: Respect rate limit specified in the kilo code settings

Ralph-Abejuela opened the pull request:
Update the lastGlobalApiRequestTime after the AI api request is done to respect the user's specified rate limit.

## Context

The rate limiting starts counting before an api request is sent NOT when an api request is finished. This causes the rate limit being inconsistent and hit rate limit on LLM models. 

Normally, API rate limits start counting when the stream has been closed. The current implementation does not do this. Which causes the rate limit specified to be violated and opens a possibility of the endpoint to be spammed. 


## Implementation

On src/core/task/Task.ts on line 2743 it does the counting before the request started. When calculating the rate limit it reduces the rate limit specified by the user by the time passed (line 2719). So this is the part where the rate limit isn't honored by kilo code. However the time pass, most of the time, is the time where the AI is generating its response. To fix this, we set the lastGlobalApiRequestTime before the request (the current code) and finally after the iterator is finished (line 2991)

so even if the request failed we got the first initial time (line 2743) and users won't wait that long and if it does succeed we update the request time so that the user's specified rate limit is respected

The fix is now toggleable with a checkbox within the advance settings

## Screenshots

N/A

## How to Test

### Normal rate limit test
- Set the rate limit to a large number (i.e. 60s)
- Do a long task (i.e ask the AI to refactor a function)
- The rate limiting should start at ~60s (i.e Rate limiting for 59s seconds...)


### Pausing rate limit test
- Set the rate limit to a large number (i.e. 60s)
- Do a long task (i.e ask the AI to refactor a function)
- Wait until "Rate limiting for 30s seconds..."
- Press cancel
- Wait for 10s
- Rate limiting should show "Rate limiting for 20s seconds..."

I've tested on my local machine and it does what it's expected

## Get in Touch

<!-- We'd love to have a way to chat with you about your changes if necessary. If you're in the [Kilo Code Discord](https://kilocode.ai/discord), please share your handle here. -->
My discord is @kaikideshuu
I also have raised a thread on discord [link](https://discord.com/channels/1349288496988160052/1425204312090480651)

Ralph-Abejuela committed:
Update the lastGlobalApiRequestTime after the AI api request is done to respect the user's specified rate limit.

Ralph-Abejuela committed:
Merge branch 'main' into RateLimitMinorFix

Ralph-Abejuela committed:
Added a new setting of rateLimitAfter which enable/disable recalculation of Task.lastGlobalApiRequestTime after the API stream has ended

chrarnoldus committed:
Apply suggestions from code review

chrarnoldus committed:
Changeset

chrarnoldus committed:
Merge branch 'main' into RateLimitMinorFix

chrarnoldus commented on packages/types/src/global-settings.ts:
```diff
@@ -139,6 +139,7 @@ export const globalSettingsSchema = z.object({
 	diagnosticsEnabled: z.boolean().optional(),
 
 	rateLimitSeconds: z.number().optional(),
+	rateLimitAfter: z.boolean().optional(), // kilocode_change
```
I suspect rateLimitSeconds is also wrong

Ralph-Abejuela committed:
Remove rateLimitAfter from global settings

Removed optional rateLimitAfter boolean from global settings as it was unnecessary for the fix.

chrarnoldus committed:
Add kilocode_change markers
